### ISSUE TITLE
parser duplicated matches not optimized out

### ISSUE THREAD

**Issue by zma-barefoot** [p4lang/p4c#2004]
As title says, duplicated matches 0x0800 are not optimized out.

```
header_type h_t {
    fields {
        f : 16;
    }
}

header h_t a;
header h_t b;

parser start {
    extract(a);
    return select(latest.f) {
        0x0800         : parse_b;
        0x0800         : parse_b;
        default        : ingress;
    }
}

parser parse_b {
    extract(b);
    return ingress;
}

control ingress { }
```

`FrontEndLast:`

```
    @name(".start") state start {
        packet.extract<h_t>(hdr.a);
        transition select(hdr.a.f) {
            16w0x800: parse_b;
            16w0x800: parse_b;
            default: accept;
        }
    }
```

It would also be nice to give a warning of duplicated matches.

**Comment by mbudiu-vmw** [comment]
I general it's not easy to tell whether the sets of two labels intersect or not. Moreover, when keysets include value_sets, this cannot even be determined statically. We could have some heuristics, but a complete solution is not very simple.

**Comment by zma-barefoot** [comment]
I agree with value sets we cannot determine duplicated matches statically.

With constant matches, can you give an example that we cannot trivial determine sets of two labels intersect or not?

**Comment by mbudiu-vmw** [comment]
The problem in general is not checking whether two labels overlap, but if label M is included in the union of the prior M-1 labels. That's not trivial for a mix of range and don't care queries, e.g., `5...8` and `6 &&& 4`

Note that the user may want overlapping labels for convenience; the only clear error case is when a label is completely redundant.

**Comment by zma-barefoot** [comment]
Okay, I see. Sounds like the general solution is to expand all labels and compute set inclusion based on ternary match, which I agree may not be trivial.

**Comment by jafingerhut** [comment]
It is an NP-complete problem, equivalent to satisfiability, and solvable using off-the-shelf SAT solver software, to determine whether rule M is matchable, vs. it is unmatchable because it is completely "covered" by the union of the previous M-1 rules, if they can be ternary.

**Comment by zma-barefoot** [comment]
> It is an NP-complete problem, equivalent to satisfiability

I'm genuinely intrigued to a see the reduction to boolean satisfiability.

**Comment by jafingerhut** [comment]
The way I first learned about the satisfiability problem was in the conjunctive normal form (CNF), i.e you have a boolean formula with one or more clauses.  Each clauses is the logical OR of boolean variables, or their negation.  The entire formula is the logical AND of all of the clauses.  The satisfiability problem asks if there is an assignment of true/false values to the variables that makes the formula true.

To transform that into a problem of whether a parser transition select statement key expression can be matched or not, make the select expression be a bit vector with one bit per boolean variable of the SAT problem.

If you have a clause like (x_1 OR !x_2 OR x_3), when there are 5 variables x_1 through x_5, turn that in to a ternary key expression that matches on exactly the opposite of that condition, i.e. (!x_1 AND x_2 AND !x_3), where x_4 and x_5 are wildcard because you do not care what their values are.  That is the condition for matching that key expression, so the original clause is the condition for _not_ matching that key expression.

So if you have one P4 key expression line per clause, each one representing the opposite condition of one clause, then the problem of determining whether a default match-anything key expression at the end can be matched or not, is equivalent to the satisfiability problem.  If you can find a set of boolean variable assignments that makes the formula true, then it will not match any of the earlier key set expressions, and it must match the default one.  If you cannot find such an assignment, then the default rule can never be matched.

**Comment by hesingh** [comment]
> The problem in general is not checking whether two labels overlap, but if label M is included in the union of the prior M-1 labels. That's not trivial for a mix of range and don't care queries, e.g., `5..8` and `6 &&& 4`
> 
> Note that the user may want overlapping labels for convenience; the only clear error case is when a label is completely redundant.

First, when this note was added p4c didn't even support the parser Range operation.  Second, now that p4c supports Range by converting the Range to ternary operations, a mix of Range and ternary operations can be checked for overlap.  One catch is that the Range to ternary (&&&) conversion is in the p4c midend as a pass.  One may want such checks in then p4c frontend.

**Comment by jafingerhut** [comment]
If the p4c developers would like to put a SAT solver library into p4c, then it is some chunk of work to be very precise and certain that you catch all possible overlaps, and warn about them all.  I have my doubts that people want to do this in p4c for every multi-line select expression.

If someone wants to create a warning when two key expressions are exactly the same as each other, as "that clearly looks like a mistake", that does not require a SAT solver to do it.  How useful it would be, not sure.  It sounds more like a "P4 lint" kind of check/warning to me, personally.  And in the lint tool you could go ahead and do the SAT solver version.

**Comment by hesingh** [comment]
In p4-16 grammar we have `keysetExpression`  which includes a simpleKeySetExpression

https://github.com/p4lang/p4c/blob/master/frontends/parsers/p4/p4parser.ypp#L771

I believe all types of allowed simpleKeysetExpression can be checked for overlap without using a SAT solver. 

The other type of `keysetExpression` is ` tupleKeysetExpression` which is a vector of 'simpleKeysetExpression`.  Again the vector can be matched for overlap - p4c implementation uses C++ and supports looping over a vector.

**Comment by jafingerhut** [comment]
Overlap is easy to check for, agreed, but I do not think overlap should generate a warning.  I suspect it is _common_ to have some overlap _on purpose_ when there is ternary matching going on, e.g. an earlier more specific rule for a special case, and a later rule that includes all of the earlier ones, but more, but you want the special case handled first.  If that generated a warning, I think people would disable/filter out/complain/learn to ignore warnings.

**Comment by hesingh** [comment]
Agree.  

Do we have to support catching identical Range and ternary operations?  If yes, this is possible as well.

**Comment by jafingerhut** [comment]
There is no "have to" here, at least not as far as the language spec is concerned.  I someone writes a parser transition that can be proved statically to be impossible to take, it makes sense to catch at least some of those.  My main concern with catching _all_ of them is that it takes a SAT solver to do it, in general.  If two transitions are identical, or a later ternary or exact entry is a subset of a _single_ earlier entry, that can be checked in O(N^2) time, where N is the number of transitions in the select expression, and N is usually pretty small in P4 programs.

**Comment by hesingh** [comment]
Note a recent PR https://github.com/p4lang/p4c/pull/2198/files does not allow duplicate parser value set (pvs) use in the parser.   I agree about the general SAT solver, but the compiler only supports a simpleKeysetExpression and if range is converted to ternary (&&&) ops, p4c can check if M is included in the prior union of M-1 labels.  If a range is converted to, say, 3 ternary op entries, then, I think, if any of the three exists in M-1 labels, it's a problem to raise an error for.

**Comment by jafingerhut** [comment]
If you want to create a check that a single simpleKeySetExpression is a subset of _any one_ of the preceding M-1 simpleKeySetExpression, that takes O(M) time, and checking all pairs takes O(N^2) time for a list of N simpleKeySetExpressions.

If you want to check whether a single simpleKeySetExpression is a subset of _the union of_ the preceding M-1 simpleKeySetExpression, then you are solving SAT to do that, in general.

**Comment by jafingerhut** [comment]
Here is an example that I would prefer no warning for, where an earlier one is a subset of a later one:
```
    0x00ff : next_state_1;
    0x00fe .. 0x00110 : next_state_2;
```
Could that be rewritten so there was no overlap?  Sure, it could, but it would be longer code, and perhaps less clear as a result.  If there were N individual overlapping values before the range, requiring a programmer to break it up into N+1 ranges to avoid a warning seems like busy-work for the programmer.

**Comment by DarinkaNesovic** [comment]
I have some questions regarding this problem.
We have next code:
    3 .. 5   : some_state0;
    1 .. 6   : some_state1;
    4 .. 10 : some_state2;
    4	       : some_state3;

Here we have a several situations that can be interpreted as error:
There are overlap of second and third label (also first and third), and first label is subset of second.
Fourth label (4) is in previous three labels, so what value she have? 
What cases are considered as error when label is constant and it is repeated in some other labels whose type is range or mask?

In midend, when ranges are replaced with masks, labels look like this:
    16w3 &&& 16w65535: some_state0;
    16w4 &&& 16w65534: some_state0;

    16w1 &&& 16w65535: some_state1;
    16w2 &&& 16w65534: some_state1;
    16w4 &&& 16w65534: some_state1;
    16w6 &&& 16w65535: some_state1;

    16w4 &&& 16w65532: some_state2;
    16w8 &&& 16w65534: some_state2;
    16w10 &&& 16w65535: some_state2;	
   
    16w4: some_state3; 

As can be seen here, in midend, from different ranges we get same masks.
What of above should be considered as error and what should be accepted?

**Comment by hesingh** [comment]
Right now, neither the midend  nor the bmv2 backend support a check for overlapping keys.

**Comment by mbudiu-vmw** [comment]
What do you mean by "error"?
The spec allows arbitrary labels and evaluates them from top to bottom. It is fine for labels to describe overlapping sets or for labels to never be matched dynamically. Some cases could be optimized away, but as you saw in the discussion, the general problem is hard.
